client/authority-discovery/src/service.rs

```diff
@@ -44,11 +44,7 @@ impl Service {
 	/// [`crate::Worker`] failed.
 	///
 	/// [`Multiaddr`]s returned always include a [`PeerId`] via a
-	/// [`libp2p::core::multiaddr:Protocol::P2p`] component. [`Multiaddr`]s
-	/// might differ in their [`PeerId`], e.g. when each [`Multiaddr`]
-	/// represents a different sentry node. This might change once support for
-	/// sentry nodes is removed (see
-	/// https://github.com/paritytech/substrate/issues/6845).
+	/// [`libp2p::core::multiaddr:Protocol::P2p`] component.
 	pub async fn get_addresses_by_authority_id(&mut self, authority: AuthorityId) -> Option<Vec<Multiaddr>> {
```

Comment:
When should we change the return type from `Option<Vec<Multiaddr>>` to `Option<Multiaddr>`?
I assume the only reason for a `Vec` was sentry nodes?

Reply:
I am afraid that the `Option<Vec<Multiaddr>>` won't go away any time soon. A libp2p node can have multiple external addresses. Some are specified by the user (e.g. by listening on specific addresses), some are reported by our peers. There is no way to tell which one is the authoritative one, if any.

Say for example a node is running behind a proxy. The node itself has IP address `2001:db8:0:0:0:0:0:1`. Nodes in the same subnet can reach our node with this address. The reversed proxy has `2001:db8:0:0:0:0:0:2`. Nodes outside the local subnet can reach our node via the proxy. Our node should advertise both addresses to allow nodes from the same subnet as well as nodes from other subnets to reach out to it. It will be made aware of the `2001:db8:0:0:0:0:0:2`address by nodes from other subnets via the [libp2p identify protocol](https://github.com/libp2p/specs/tree/master/identify).

Reply:
Thanks for the clarification. So then the `PeerId` component of each `Multiaddr` would be the same? I would add this in place of removed doc comments.

Reply:
> So then the PeerId component of each Multiaddr would be the same?

Yes, had the same thought initially. Problem is that for now there are still sentry nodes out there, thus at least during the transition period we can not provide this promise.

Reply:
I would still be in favor of adding some comments about that (maybe not doc comment but just regular comments).

Reply:
@ordian what do you think of https://github.com/paritytech/substrate/pull/7368/commits/f0071a1bb1df9fe1472504e8e847e26f2bd13afa detailing the background and the future possibility?

Reply:
Thanks, looks good!